buildSrc/src/main/kotlin/pklJavaLibrary.gradle.kts

```diff
@@ -1,27 +1,51 @@
 @file:Suppress("HttpUrlsUsage")
 
+import org.gradle.accessors.dm.LibrariesForLibs
+
 plugins {
+  pmd
   `java-library`
   id("pklKotlinTest")
   id("com.diffplug.spotless")
 }
 
-// make sources Jar available to other subprojects
-val sourcesJarConfiguration = configurations.register("sourcesJar")
+// JVM toolchain defaults, properties, and resolved configuration.
+private val defaultJvmTarget = "11"
+private val jvmVendor = JvmVendorSpec.GRAAL_VM
+private val jvmTargetVersion =
+  (findProperty("javaTarget") as? String ?: defaultJvmTarget)
+
+// Version catalog library symbols.
+val libs = the<LibrariesForLibs>()
+
+// make source jar available to other subprojects
+val sourcesJarConfiguration: Provider<Configuration> = configurations.register("sourcesJar")
+
+pmd {
+  isConsoleOutput = true
+  toolVersion = libs.versions.pmd.get()
+  threads = 4
+  isIgnoreFailures = true
+  incrementalAnalysis = true
+}
 
 java {
+  // obtain and use a Java toolchain from GraalVM, at the version specified for the project.
+  sourceCompatibility = JavaVersion.toVersion(jvmTargetVersion)
+  targetCompatibility = JavaVersion.toVersion(jvmTargetVersion)
+
   withSourcesJar() // creates `sourcesJar` task
   withJavadocJar()
 }
 
 artifacts {
-  // make sources Jar available to other subprojects
+  // make source jar available to other subprojects
   add("sourcesJar", tasks["sourcesJar"])
 }
 
 spotless {
   java {
-    googleJavaFormat("1.15.0")
+    googleJavaFormat(libs.versions.googleJavaFormat.get())
```

Comment:
All tools are now listed in the Version Catalog, and a fix was applied to make the version catalog accessible from precompiled Kotlin convention plugins.